apps/hash-ai-worker-ts/src/activities/shared/find-existing-entity.ts

```diff
@@ -0,0 +1,291 @@
+  proposedEntity,
+}: {
+  actorId: AccountId;
+  dereferencedEntityType?: DereferencedEntityType;
+  graphApiClient: GraphApi;
+  ownedById: OwnedById;
+  proposedEntity: Pick<ProposedEntity, "entityTypeId" | "properties">;
+}): Promise<Entity | undefined> => {
+  const entityTypeId = proposedEntity.entityTypeId;
+
+  const entityType: DereferencedEntityType | undefined =
+    dereferencedEntityType ??
+    (await graphApiClient
+      .getEntityTypesByQuery(actorId, {
+        includeDrafts: false,
+
+        filter: {
+          equal: [{ path: ["versionedUrl"] }, { parameter: entityTypeId }],
+        },
+        graphResolveDepths: {
+          ...zeroedGraphResolveDepths,
+          ...fullOntologyResolveDepths,
+        },
+        temporalAxes: currentTimeInstantTemporalAxes,
+      })
+      .then(({ data }) => {
+        const subgraph = mapGraphApiSubgraphToSubgraph<EntityTypeRootType>(
+          data,
+          actorId,
+        );
+
+        const foundEntityType = getRoots(subgraph)[0];
+
+        if (!foundEntityType) {
+          return;
+        }
+
+        return dereferenceEntityType(entityTypeId, subgraph).schema;
+      }));
+
+  if (!entityType) {
+    throw new Error(`Could not retrieve EntityType with id ${entityTypeId}`);
+  }
+
+  /**
+   * Search for an existing entity that seems to semantically match the proposed entity, based on:
+   * 1. The value for the label property, or other properties which are good candidates for unique identifying
+   * an entity
+   * 2. The entire entity properties object, if there is no match from (1)
+   */
+
+  /** We are going to need the embeddings in both cases, so create these first */
+  const { embeddings } = await createEntityEmbeddings({
+    entityProperties: proposedEntity.properties,
+    propertyTypes: Object.values(entityType.properties).map(
+      (propertySchema) => ({
+        title:
+          "items" in propertySchema
+            ? propertySchema.items.title
+            : propertySchema.title,
+        $id:
+          "items" in propertySchema
+            ? propertySchema.items.$id
+            : propertySchema.$id,
+      }),
+    ),
+  });
+
+  const existingEntityBaseAllFilter = [
+    { equal: [{ path: ["archived"] }, { parameter: false }] },
+    {
+      equal: [
+        { path: ["ownedById"] },
+        {
+          parameter: ownedById,
+        },
+      ],
+    },
+    generateVersionedUrlMatchingFilter(entityTypeId),
+  ] satisfies AllFilter["all"];
+
+  const maximumSemanticDistance = 0.55;
+
+  /**
+   * First find suitable specific properties to match on
+   */
+  const labelProperty = entityType.labelProperty;
+
+  const propertyBaseUrlsToMatchOn: BaseUrl[] =
+    labelProperty &&
+    entityType.properties[labelProperty] &&
+    proposedEntity.properties[labelProperty]
+      ? [labelProperty]
+      : [];
+
+  const nameProperties = [
+    "name",
+    "legal name",
+    "preferred name",
+    "profile url",
+    "shortname",
+  ];
```

Comment:
I've added "organization name" in https://github.com/hashintel/hash/pull/4298/commits/3ca002c179495acf2f9033d7a4c9d0c3dffc41b8

I don't think we should unify it with the one in `generateEntityLabel`, because that is just looking for something that might be a label for a single entity, and it matters much less if some other entity of the same type has the same label. Whereas here, we are taking the value to mean 'this is the same entity', so for things like `title` (Mr, Mrs, Senior Engineer, etc) and to a lesser extent `display name` we increase the risk of false positives. This is also true of "preferred name" and "name", for what it's worth. 

I expect we'll need to improve how this matching works after further testing, as I've already seen a few false positives in the limited number of entities we're dealing with. Ideally being able to identify which properties uniquely identify an entity of a given type.